File: client/src/mocaclient/Util/MocaType.cs
using System;
using System.Data;
using System.Diagnostics;

namespace RedPrairie.MOCA.Client
{
    /// <summary>
    /// Enumeration class to define the known data types that can be
    /// a part of a MOCA result set. 
    /// </summary>
    [Serializable]
    public class MocaType
    {
        /// <summary>
        /// A static moca type that defines an integer column implimented by <see cref="Int32"/>.
        /// </summary>
        public static readonly MocaType INTEGER = new MocaType('I', typeof (Int32), DbType.Int32, "INTEGER");
        /// <summary>
        /// A static moca type that defines an Integer reference column implimented by <see cref="Int32"/>.
        /// </summary>
        public static readonly MocaType INTEGER_REF = new MocaType('P', typeof (Int32), DbType.Int32, "INTEGER_REF");
        /// <summary>
        /// A static moca type that defines a double column implimented by <see cref="Double"/>.
        /// </summary>
        public static readonly MocaType DOUBLE = new MocaType('F', typeof (Double), DbType.Double, "DOUBLE");
        /// <summary>
        /// A static moca type that defines a double reference column implimented by <see cref="Double"/>.
        /// </summary>
        public static readonly MocaType DOUBLE_REF = new MocaType('X', typeof (Double), DbType.Double, "DOUBLE_REF");
        /// <summary>
        /// A static moca type that defines a string column implimented by <see cref="String"/>.
        /// </summary>
        public static readonly MocaType STRING = new MocaType('S', typeof (string), DbType.String, "STRING");
        /// <summary>
        /// A static moca type that defines a string reference column implimented by <see cref="String"/>.
        /// </summary>
        public static readonly MocaType STRING_REF = new MocaType('Z', typeof (string), DbType.String, "STRING_REF");
        /// <summary>
        /// A static moca type that defines a Date/Time column implimented by <see cref="DateTime"/>.
        /// </summary>
        public static readonly MocaType DATETIME = new MocaType('D', typeof (DateTime), DbType.DateTime, "DATETIME");
        /// <summary>
        /// A static moca type that defines a Boolean column implimented by <see cref="Boolean"/>.
        /// </summary>
        public static readonly MocaType BOOLEAN = new MocaType('O', typeof (Boolean), DbType.Boolean, "BOOLEAN");
        /// <summary>
        /// A static moca type that defines a binary column implimented by a <see cref="byte"/> array.
        /// </summary>
        public static readonly MocaType BINARY = new MocaType('V', typeof (byte[]), DbType.Binary, "BINARY");
        /// <summary>
        /// A static moca type that defines another Moca results column.
        /// </summary>
        public static readonly MocaType RESULTS;
        /// <summary>
        /// A static moca type that defines a generic object column implimented by an <see cref="object"/>.
        /// </summary>
        public static readonly MocaType OBJECT = new MocaType('J', typeof (Object), DbType.Object, "OBJ");
        /// <summary>
        /// A static moca type that defines an unknown column type.
        /// </summary>
        public static readonly MocaType UNKNOWN = new MocaType('?', typeof (Object), DbType.Object, "UNKNOWN");

        #region Private Fields

        private readonly char _code;
        [NonSerialized] private Type _valueClass;
        [NonSerialized] private DbType _sqlType;
        [NonSerialized] private string _text;

        #endregion

        #region Constructors

        /// <summary>
        /// Private Constructor
        /// </summary>
        /// <param name="code">The MOCA Type code</param>
        /// <param name="cls">The system data type</param>
        /// <param name="sqlType">The <see cref="DbType"/> </param>
        /// <param name="text">A description</param>
        private MocaType(char code, Type cls, DbType sqlType, string text)
        {
            _code = code;
            _valueClass = cls;
            _sqlType = sqlType;
            _text = text;
        }

        [DebuggerStepThrough]
        static MocaType()
        {
            RESULTS = new MocaType('R', typeof (DataTable), DbType.Object, "RESULTS");
        }

        #endregion

        #region Public Properties

        /// <summary> 
        /// Returns the MOCA type code (COMTYP_*) corresponding to this type.
        /// </summary>
        /// <returns> the MOCA type code for this type. </returns>
        public char TypeCode
        {
            get { return _code; }
        }

        /// <summary> 
        /// Returns the system type that represents this type.
        /// </summary>
        /// <returns> the system type that represents this type. </returns>
        public Type Class
        {
            get { return _valueClass; }
        }

        /// <summary> 
        /// Returns the <see cref="DbType"/> of the given MOCA type.  This is mainly useful
        /// for use with <see cref="DataSet"/> manipulation
        /// </summary>
        /// <returns> the <see cref="DbType"/> corresponding to this data type. </returns>
        public DbType SQLType
        {
            get { return _sqlType; }
        }

        #endregion

        #region Public Methods

        /// <summary> 
        /// Returns the <code>MocaType</code> corresponding to the given low-level
        /// MOCA type code.
        /// </summary>
        /// <param name="code">a MOCA type code (COMTYP_*).</param>
        /// <returns> the corresponding instance of this enumerated class, or
        /// <code>UNKNOWN</code> if the type code is unrecognized.
        /// </returns>
        public static MocaType Lookup(char code)
        {
            switch (code)
            {
                case 'I':
                    return INTEGER;

                case 'L':
                    return INTEGER;

                case 'P':
                    return INTEGER_REF;

                case 'F':
                    return DOUBLE;

                case 'X':
                    return DOUBLE_REF;

                case 'S':
                    return STRING;

                case 'Z':
                    return STRING_REF;

                case 'D':
                    return DATETIME;

                case 'O':
                    return BOOLEAN;

                case 'V':
                    return BINARY;

                case 'R':
                    return RESULTS;

                case 'J':
                    return OBJECT;

                default:
                    return UNKNOWN;
            }
        }

        /// <summary> 
        /// Determine what MOCA type is appropriate for the given class.  This
        /// method will never return reference types, as they are indistinguishable
        /// from their value counterparts.
        /// </summary>
        /// <param name="cls">the class to be used to look up a MOCA type object. </param>
        /// <returns> the <code>MocaType</code> object corresponding to the given
        /// class.  If there is no known type for <code>cls</code>,
        /// <code>UNKNOWN</code> is returned.
        /// </returns>
        public static MocaType LookupClass(Type cls)
        {
            if (cls.Equals(typeof (string)))
                return STRING;
            if (cls.Equals(typeof (Int32)) || cls.Equals(Type.GetType("System.Int32")))
                return INTEGER;
            if (cls.Equals(typeof (Double)) || cls.Equals(Type.GetType("System.Double")))
                return DOUBLE;
            if (cls.Equals(typeof (Boolean)) || cls.Equals(Type.GetType("System.Boolean")))
                return BOOLEAN;
            if (typeof (DateTime).IsAssignableFrom(cls))
                return DATETIME;
            if (typeof (ValueType).IsAssignableFrom(cls))
                return DOUBLE;
            if (cls.Equals(typeof (DataTable)))
                return RESULTS;
            if (cls.IsArray && cls.GetElementType().Equals(Type.GetType("System.Byte")))
                return BINARY;
            if (!cls.IsPrimitive && !cls.IsArray)
                return OBJECT;
            
            return UNKNOWN;
        }

        #endregion

        #region Public Method Overrides

        /// <summary>
        /// Determines if an object is equal to the current object.
        /// </summary>
        /// <param name="o">The object to compare.</param>
        /// <returns><c>true</c> if they are the same; otherwise <c>false</c>.</returns>
        public override bool Equals(Object o)
        {
            if (ReferenceEquals(this, o)) return true;

            MocaType mType = (MocaType) o;

            if (mType != null)
            {
                return (_code == mType._code);
            }
            
            return false;
        }

        /// <summary>
        /// Serves as a hash function for a particular type. <see cref="M:System.Object.GetHashCode"></see> is suitable for use in hashing algorithms and data structures like a hash table.
        /// </summary>
        /// <returns>
        /// A hash code for the current <see cref="T:System.Object"></see>.
        /// </returns>
        public override int GetHashCode()
        {
            return _code;
        }

        /// <summary>
        /// Returns a <see cref="T:System.String"></see> that represents the current <see cref="T:System.Object"></see>.
        /// </summary>
        /// <returns>
        /// A <see cref="T:System.String"></see> that represents the current <see cref="T:System.Object"></see>.
        /// </returns>
        public override string ToString()
        {
            return _text;
        }

        /// <summary>
        /// Implements the operator ==.
        /// </summary>
        /// <param name="a">The first object to compare.</param>
        /// <param name="b">The second object to compare.</param>
        /// <returns>The result of the operator.</returns>
        public static bool operator ==(MocaType a, MocaType b)
        {
            if (ReferenceEquals(a, b)) return true;

            if ((object) a == null && (object) b == null) return true;

            if ((object) a == null || (object) b == null) return false;

            return a.Equals(b);
        }

        /// <summary>
        /// Implements the operator !=.
        /// </summary>
        /// <param name="a">The first object to compare.</param>
        /// <param name="b">The second object to compare.</param>
        /// <returns>The result of the operator.</returns>
        public static bool operator !=(MocaType a, MocaType b)
        {
            return !(a == b);
        }

        #endregion
    }
}

File: src/main/java/core/com/redprairie/moca/MocaType.java
/*
 *  $URL$
 *  $Revision$
 *  $Author$
 *  $Date$
 *  
 *  $Copyright-Start$
 *
 *  Copyright (c) 2016
 *  Sam Corporation
 *  All Rights Reserved
 *
 *  This software is furnished under a corporate license for use on a
 *  single computer system and can be copied (with inclusion of the
 *  above copyright) only for use on such a system.
 *
 *  The information in this document is subject to change without notice
 *  and should not be construed as a commitment by Sam Corporation.
 *
 *  Sam Corporation assumes no responsibility for the use of the
 *  software described in this document on equipment which has not been
 *  supplied or approved by Sam Corporation.
 *
 *  $Copyright-End$
 */

package com.redprairie.moca;

import java.sql.Types;
import java.util.Date;

import com.redprairie.moca.server.legacy.GenericPointer;

/**
 * Enumeration class to define the known data types that can be
 * a part of a MOCA result set. 
 * 
 * <b><pre>
 * Copyright (c) 2016 Sam Corporation
 * All rights reserved.
 * </pre></b>
 *
 * @author  Derek Inksetter
 * @version $Revision$
 */
public enum MocaType
{
    // Order is important, for type promotion.
    // BOOLEAN < INTEGER < DOUBLE < DATETIME < STRING
    // All others are not significant.
    BOOLEAN('O', Boolean.class, Types.BOOLEAN),
    INTEGER('I', Integer.class, Types.INTEGER),
    INTEGER_REF('P', Integer.class, Types.INTEGER),
    DOUBLE('F', Double.class, Types.DOUBLE),
    DOUBLE_REF('X', Double.class, Types.DOUBLE),
    DATETIME('D', Date.class, Types.TIMESTAMP),
    STRING('S', String.class, Types.VARCHAR),
    STRING_REF('Z', String.class, Types.VARCHAR),
    BINARY('V', byte[].class, Types.BINARY),
    RESULTS('R', MocaResults.class, Types.OTHER),
    OBJECT('J', Object.class, Types.OTHER),
    GENERIC('G', GenericPointer.class, Types.OTHER),
    UNKNOWN('?', Object.class, Types.OTHER);
    
    /**
     * Returns the MOCA type code (COMTYP_*) corresponding to this type.
     * @return the MOCA type code for this type.
     */
    public char getTypeCode() {
    	return _code;
    }
    
    /**
     * Returns the Java class that represents this type.
     * @return the Java class that represents this type.
     */
    public Class<? extends Object> getValueClass() {
        return _valueClass;
    }
    
    /**
     * Returns the JDBC type of the given MOCA type.  This is mainly useful
     * for use with MOCA's BindList class.
     * @return the JDBC type code corresponding to this data type.
     */
    public int getSQLType() {
        return _sqlType;
    }
    
    /**
     * Returns the <code>MocaType</code> corresponding to the given low-level
     * MOCA type code.
     * @param code a MOCA type code (COMTYP_*).
     * @return the corresponding instance of this enumerated class, or
     * <code>UNKNOWN</code> if the type code is unrecognized.
     */
    public static MocaType lookup(char code) {
        switch (code) {
            case 'I': 
            case 'L': return INTEGER;
            case 'P': return INTEGER_REF;
            case 'F': return DOUBLE;
            case 'X': return DOUBLE_REF;
            case 'S': return STRING;
            case 'Z': return STRING_REF;
            case 'D': return DATETIME;
            case 'O': return BOOLEAN;
            case 'V': return BINARY;
            case 'R': return RESULTS;
            case 'J': return OBJECT;
            case 'G': return GENERIC;
            default: return UNKNOWN;
        }
    }
    
    /**
     * Determine what MOCA type is appropriate for the given class.  This
     * method will never return reference types, as they are indistinguishable
     * from their value counterparts.
     * @param cls the class to be used to look up a MOCA type object.
     * @return the <code>MocaType</code> object corresponding to the given
     * class.  If there is no known type for <code>cls</code>,
     * <code>UNKNOWN</code> is returned.
     */
    public static MocaType lookupClass(Class<?> cls) {
        if (cls.equals(String.class))
            return STRING;
        else if (cls.equals(Integer.class) || cls.equals(Integer.TYPE))
            return INTEGER;
        else if (cls.equals(Double.class) || cls.equals(Double.TYPE))
            return DOUBLE;
        else if (cls.equals(Boolean.class) || cls.equals(Boolean.TYPE))
            return BOOLEAN;
        else if (Date.class.isAssignableFrom(cls))
            return DATETIME;
        else if (cls.equals(GenericPointer.class))
            return GENERIC;
        else if (MocaResults.class.isAssignableFrom(cls))
            return RESULTS;
        else if (Number.class.isAssignableFrom(cls))
            return DOUBLE;
        else if (cls.isArray() && cls.getComponentType().equals(Byte.TYPE))
            return BINARY;
        else if (cls.equals(GenericPointer.class))
            return GENERIC;
        else if (!cls.isPrimitive() && !cls.isArray())
            return OBJECT;
        else
            return UNKNOWN;
    }
    
    public static MocaType forValue(Object value) {
        if (value == null) {
            return MocaType.STRING;
        }
        else {
            return lookupClass(value.getClass());
        }
    }
    
    //
    // Implementation
    //
    
    /**
     * Private constructor for data types.
     * @param code
     * @param cls
     */
    private MocaType(char code, Class<? extends Object> cls, int sqlType) {
        _code = code;
        _valueClass = cls;
        _sqlType = sqlType;
    }
    
    private static final long serialVersionUID = 3834305120656962609L;
    private final char _code;
    private final transient Class<? extends Object> _valueClass;
    private final transient int _sqlType;
}
